### DISCUSSION TITLE
How to use r2d2 with diesel_async?

### DISCUSSION THREAD

**Discussion by Abdillah** [weiznich/diesel_async#113] [Q&A]
How to use r2d2 with diesel_async? As I saw the feature here.

I do this but found an error on below.

```
let db_factory = std::thread::spawn(|| {
    let mut rt = tokio::runtime::Runtime::new().unwrap();
    let local = tokio::task::LocalSet::new();
    let url = dotenv::var("DB_CONNSTR").unwrap();
    local.block_on(&mut rt, establish_connection(&url)).unwrap()
});
println!("Creating database connection..");
let db = diesel_async::pooled_connection::AsyncDieselConnectionManager::new(
    db_factory.join().expect("Thread db_factory panicked"),
);
let pool = diesel::r2d2::Pool::builder()
    .max_size(15)
    .build(db)
    .unwrap();
```

```
error[E0277]: the trait bound `AsyncDieselConnectionManager<AsyncPgConnection>: ManageConnection` is not satisfied
   --> src/main.rs:65:9
    |
65  | /         diesel::r2d2::Pool<
66  | |             diesel_async::pooled_connection::AsyncDieselConnectionManager<AsyncPgConnection>,
67  | |         >,
    | |_________^ the trait `ManageConnection` is not implemented for `AsyncDieselConnectionManager<AsyncPgConnection>`
    |
```

**Comment by weiznich** [comment]
It's not supported to use `r2d2` directly with any `AsyncConnection` type provided by diesel-async, as using an sync connection pool with an async connection is an anti-pattern. You either want to use an async-connection pool (like `bb8`, `deadpool` or `mobc`) with an async connection type or you want to use a sync connection pool (like `r2d2`) with a sync connection type (like `diesel::PgConnection`, or `diesel_async::async_connection_wrapper::AsyncConnectionWrapper<diesel_async::AsyncPgConnection>`)

**Reply by Abdillah** [reply]
Ah, okay thanks.. 
I tried to implement it myself, but turns out I need to impl `diesel::Connection` to `AsyncPgConnection` which is not possible due to conflicting methods.

**Reply by weiznich** [reply]
You might want to use `diesel_async::async_connection_wrapper::AsyncConnectionWrapper<diesel_async::AsyncPgConnection>` in that case. It provides such an implementation.
